tests/unit/azure/test_credentials.py

```diff
@@ -0,0 +1,337 @@
+    assert sp_1.permission_mapping.principal == storage_credential.name
+    assert storage_credential.read_only is False
+
+    storage_credential = credential_manager.create_storage_credential(sp_2)
+    assert sp_2.permission_mapping.principal == storage_credential.name
+    assert storage_credential.read_only is True
+
+
+def test_validate_storage_credentials(credential_manager):
+    sp_1 = ServicePrincipalMigrationInfo(
+        StoragePermissionMapping(
+            "prefix1",
+            "app_secret1",
+            "principal_1",
+            "WRITE_FILES",
+            "directory_id_1",
+        ),
+        "test",
+    )
+    sc_1 = StorageCredentialInfo(
+        name=sp_1.permission_mapping.principal,
+        azure_service_principal=AzureServicePrincipal(
+            sp_1.permission_mapping.directory_id, sp_1.permission_mapping.client_id, sp_1.client_secret
+        ),
+        read_only=False,
+    )
+
+    sp_2 = ServicePrincipalMigrationInfo(
+        StoragePermissionMapping(
+            "prefix2",
+            "app_secret2",
+            "principal_read",
+            "READ_FILES",
+            "directory_id_1",
+        ),
+        "test",
+    )
+    sc_2 = StorageCredentialInfo(
+        name=sp_2.permission_mapping.principal,
+        azure_service_principal=AzureServicePrincipal(
+            sp_2.permission_mapping.directory_id, sp_2.permission_mapping.client_id, sp_2.client_secret
+        ),
+        read_only=True,
+    )
+
+    sp_3 = ServicePrincipalMigrationInfo(
+        StoragePermissionMapping(
+            "overlap_with_external_location",
+            "app_secret4",
+            "principal_overlap",
+            "WRITE_FILES",
+            "directory_id_2",
+        ),
+        "test",
+    )
+    sc_3 = StorageCredentialInfo(
+        name=sp_3.permission_mapping.principal,
+        azure_service_principal=AzureServicePrincipal(
+            sp_3.permission_mapping.directory_id, sp_3.permission_mapping.client_id, sp_3.client_secret
+        ),
+    )
+
+    # validate normal storage credential
+    validation = credential_manager.validate_storage_credential(sc_1, sp_1)
+    assert validation.read_only is False
+    assert validation.name == sp_1.permission_mapping.principal
+    for result in validation.results:
+        if result.operation.value == "WRITE":
+            assert result.result.value == "PASS"
+
+    # validate read-only storage credential
+    validation = credential_manager.validate_storage_credential(sc_2, sp_2)
+    assert validation.read_only is True
+    assert validation.name == sp_2.permission_mapping.principal
+    for result in validation.results:
+        if result.operation.value == "READ":
+            assert result.result.value == "PASS"
+
+    # prefix used for validation overlaps with existing external location
+    validation = credential_manager.validate_storage_credential(sc_3, sp_3)
+    assert (
+        validation.results[0].message
```

Comment:
definitely simplify this data structure, so that tests are easier to read

Reply:
refactored